test/sqllogictest/github-8241.slt

```diff
@@ -39,3 +39,138 @@ DETAIL: The following relations in the query are outside the transaction's time
 "materialize.public.i2"
 Only the following relations are available:
 "materialize.public.i1"
+"materialize.public.t1"
```

Comment:
I didn't take a close look at this change, but I'm actually a bit baffled by this. Why wasn't `materialize.public.t1` included in the timedomain previously? It's in the same schema as `materialize.public.i1`. Also, why did my changes cause it to be included? My changes only added system schemas, it didn't touch user schemas. I'll have to look into this.

Anyway, to answer your questions directly:

> Do we have data on how expensive it is to add all these read holds?

I don't have any data on this, but it's expensiveness would depend on how long the transaction is held open.

>  Is it mostly just adding to a btreemap so it's very cheap?

My feeling is that it's rather expensive on the system indexes since compaction is stopped and they will just accumulate in-memory in the clusters.

> Also having system and user id'd indexes seems not great, as that will balloon as users add many objects across many clusters.

We only include indexes in the active cluster and the `mz_introspection` cluster. So adding indexes across many clusters won't cause the read holds of a single transaction to balloon to include all of those indexes. 

I think we have to at least sometimes mix user and system indexes. A query like `SELECT * FROM t, mz_views` already includes system indexes (the introspection sources of the current cluster) and user indexes (any indexes in the active cluster). This change just makes that the default behavior.

> One of the arguments for only including pg_catalog is that it would limit this list of included indexes to a known set that doesn't change based on user created objects.

I'm not sure I'm following this. Are you suggesting to exclude user indexes from the timedomain? That would prevent users from using user indexes in a transaction which doesn't seem ideal. Or are you suggesting to exclude the introspection sources? Either way, I think you're right about only including `pg_catalog` by default. We can re-visit later if the other schemas are needed.

Reply:
> Why wasn't materialize.public.t1 included in the timedomain previously? It's in the same schema as materialize.public.i1. Also, why did my changes cause it to be included?

Ok I figured this out. A relation will be excluded from the timedomain, if there is an index on that relation in the timedomain.